sycl/include/sycl/ext/oneapi/experimental/cuda/non_uniform_algorithms.hpp

```diff
@@ -0,0 +1,367 @@
+
+  return non_uniform_shfl(g, MemberMask, x, broadID);
+}
+
+// Non Redux types must fall back to shfl based implementations.
+template <typename Group, typename T, class BinaryOperation>
+std::enable_if_t<
+    std::is_same<IsRedux<T, BinaryOperation>, std::false_type>::value &&
+        ext::oneapi::experimental::is_user_constructed_group_v<Group>,
+    T>
+masked_reduction_cuda_sm80(Group g, T x, BinaryOperation binary_op,
+                           const uint32_t MemberMask) {
+  return masked_reduction_cuda_shfls(g, x, binary_op, MemberMask);
+}
+
+// get_identity is only currently used in this cuda specific header. If in the
+// future it has more general use it should be moved to a more appropriate
+// header.
+template <typename T, class BinaryOperation>
+inline __SYCL_ALWAYS_INLINE
+    std::enable_if_t<IsPlus<T, BinaryOperation>::value ||
+                         IsBitOR<T, BinaryOperation>::value ||
+                         IsBitXOR<T, BinaryOperation>::value,
+                     T>
+    get_identity() {
+  return 0;
+}
+
+template <typename T, class BinaryOperation>
+inline __SYCL_ALWAYS_INLINE
+    std::enable_if_t<IsMultiplies<T, BinaryOperation>::value, T>
+    get_identity() {
+  return 1;
+}
+
+template <typename T, class BinaryOperation>
+inline __SYCL_ALWAYS_INLINE
+    std::enable_if_t<IsBitAND<T, BinaryOperation>::value, T>
+    get_identity() {
+  return ~0;
+}
+
+#define GET_ID(OP_CHECK, OP)                                                   \
+  template <typename T, class BinaryOperation>                                 \
+  inline __SYCL_ALWAYS_INLINE                                                  \
+      std::enable_if_t<OP_CHECK<T, BinaryOperation>::value, T>                 \
+      get_identity() {                                                         \
+    if constexpr (std::is_same_v<T, char>) {                                   \
+      return std::numeric_limits<char>::OP();                                  \
+    } else if constexpr (std::is_same_v<T, unsigned char>) {                   \
+      return std::numeric_limits<unsigned char>::OP();                         \
+    } else if constexpr (std::is_same_v<T, short>) {                           \
+      return std::numeric_limits<short>::OP();                                 \
+    } else if constexpr (std::is_same_v<T, unsigned short>) {                  \
+      return std::numeric_limits<unsigned short>::OP();                        \
+    } else if constexpr (std::is_same_v<T, int>) {                             \
+      return std::numeric_limits<int>::OP();                                   \
+    } else if constexpr (std::is_same_v<T, unsigned int>) {                    \
+      return std::numeric_limits<unsigned int>::OP();                          \
+    } else if constexpr (std::is_same_v<T, long>) {                            \
+      return std::numeric_limits<int>::OP();                                   \
```

Comment:
Is it intentional that we return the numeric limits of `int` when asked for the ones for `long`?

Reply:
I did do this intentionally but I think it is wrong and I will correct it!

Reply:
In that case, is there a reason we can't just have it return `std::numeric_limits<T>::OP();` no matter the type, or are there types outside these checks?